File: scripts/wgsl_include.py
#!/usr/bin/env python3
"""
`#include "_lib.wgsl"` expansion for WGSL — Python mirror.

Mirror of src/wasm/bridge/wgslInclude.ts, which is the source of truth. The two
are held byte-identical on the fixture corpus by scripts/verify-wgsl-include.mjs
(`npm run verify:wgsl-include`), so a change here without a matching change
there fails CI.

Exists because scripts/bindgroup_checker.py parses raw WGSL text and would see
no bindings at all in a shader that gets them from _prelude.wgsl. The other
Python audits (extraBuffer, dead sliders, workgroup size, orphan defs) read
only the shader's own body and do not need expansion.

Contract: src/contracts/wgsl_include.json
"""

import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
SHADER_ROOT = PROJECT_ROOT / "public" / "shaders"

INCLUDE_RE = re.compile(r'^[ \t]*#include[ \t]+"([^"]+)"[ \t]*$')
HAS_INCLUDE_RE = re.compile(r"^[ \t]*#include[ \t]", re.MULTILINE)

MAX_DEPTH = 8
LIBRARY_PREFIX = "_"


class WgslIncludeError(Exception):
    """Raised for a malformed, missing, cyclic or repeated include."""


def strip_comments(source: str) -> str:
    """
    Blank out comment bodies, preserving every newline so line numbers and
    offsets stay aligned with the original.
    """
    out = []
    i = 0
    n = len(source)
    while i < n:
        two = source[i : i + 2]
        if two == "//":
            while i < n and source[i] != "\n":
                out.append(" ")
                i += 1
        elif two == "/*":
            depth = 0
            while i < n:
                pair = source[i : i + 2]
                if pair == "/*":
                    depth += 1
                    out.append("  ")
                    i += 2
                elif pair == "*/":
                    depth -= 1
                    out.append("  ")
                    i += 2
                    if depth == 0:
                        break
                else:
                    out.append("\n" if source[i] == "\n" else " ")
                    i += 1
        else:
            out.append(source[i])
            i += 1
    return "".join(out)


def has_wgsl_include(source: str) -> bool:
    """True when the source has at least one live (non-commented) directive."""
    return bool(HAS_INCLUDE_RE.search(strip_comments(source)))


def _validate_name(name: str, parent: str) -> None:
    if "/" in name or "\\" in name:
        raise WgslIncludeError(
            f'{parent}: #include "{name}" — public/shaders is flat, a path separator is not allowed'
        )
    if not name.endswith(".wgsl"):
        raise WgslIncludeError(f'{parent}: #include "{name}" — must end in .wgsl')
    if not name.startswith(LIBRARY_PREFIX):
        raise WgslIncludeError(
            f'{parent}: #include "{name}" — only "{LIBRARY_PREFIX}"-prefixed library files may be '
            "included, so a catalog shader cannot be inlined into another"
        )


def _default_resolver(name: str):
    path = SHADER_ROOT / name
    if not path.is_file():
        return None
    return path.read_text(encoding="utf-8")


def expand_wgsl_includes(source: str, resolve=None, entry: str = "<entry>") -> str:
    """
    Expand every `#include` in *source*.

    Returns *source* unchanged when it has no directive, so the overwhelming
    majority of the catalog costs one regex search.

    *resolve* takes an include name and returns its text, or None if missing.
    Defaults to reading from public/shaders.
    """
    if not has_wgsl_include(source):
        return source

    if resolve is None:
        resolve = _default_resolver

    seen = {}

    def expand(text: str, file: str, stack) -> str:
        if len(stack) > MAX_DEPTH:
            chain = " -> ".join(list(stack) + [file])
            raise WgslIncludeError(f"#include nesting deeper than {MAX_DEPTH}: {chain}")

        lines = text.split("\n")
        blanked = strip_comments(text).split("\n")
        out = []

        for i, line in enumerate(lines):
            match = INCLUDE_RE.match(blanked[i]) if i < len(blanked) else None
            if not match:
                out.append(line)
                continue

            name = match.group(1)
            line_no = i + 1
            _validate_name(name, file)

            if name in stack or name == file:
                chain = " -> ".join(list(stack) + [file, name])
                raise WgslIncludeError(f"#include cycle: {chain}")

            previous = seen.get(name)
            if previous is not None:
                raise WgslIncludeError(
                    f'{file}:{line_no}: #include "{name}" was already included by {previous} — '
                    "WGSL has no include guards, so the second copy would redeclare every symbol"
                )
            seen[name] = f"{file}:{line_no}"

            included = resolve(name)
            if included is None:
                raise WgslIncludeError(
                    f'{file}:{line_no}: #include "{name}" — file not found under public/shaders'
                )

            out.append(f"// #include-expanded: {name} (from {file}:{line_no})")
            out.append(expand(included, name, list(stack) + [file]))

        return "\n".join(out)

    return expand(source, entry, [])


def expand_file(path) -> str:
    """Expand a shader file on disk, using its basename in diagnostics."""
    path = Path(path)
    return expand_wgsl_includes(path.read_text(encoding="utf-8"), entry=path.name)


if __name__ == "__main__":
    import sys

    if len(sys.argv) != 2:
        print("usage: wgsl_include.py <shader.wgsl>", file=sys.stderr)
        raise SystemExit(2)
    try:
        print(expand_file(sys.argv[1]), end="")
    except WgslIncludeError as exc:
        print(f"error: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: src/wasm/bridge/wgslInclude.ts
/**
 * `#include "_lib.wgsl"` expansion for WGSL.
 *
 * Source of truth for the expander. Lives under src/wasm/bridge so
 * scripts/emit-wasm-bridge.mjs emits it to both bridge copies, which lets the
 * WASM load path and the WebGPU fetch path share one implementation instead of
 * growing a second parser in C++.
 *
 * Contract: src/contracts/wgsl_include.json
 * Python mirror: scripts/wgsl_include.py (kept identical by verify:wgsl-include)
 *
 * Deliberately minimal — textual substitution of whole files, nothing else.
 * No macros, no conditionals, no include guards: WGSL cannot redeclare a
 * symbol, so including a library twice is an error rather than a no-op.
 */

/** Whole-line directive. Mirrors wgsl_include.json `directive.pattern`. */
const INCLUDE_RE = /^[ \t]*#include[ \t]+"([^"]+)"[ \t]*$/;

export const WGSL_INCLUDE_MAX_DEPTH = 8;
export const WGSL_LIBRARY_PREFIX = '_';

export class WgslIncludeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'WgslIncludeError';
  }
}

/** Resolves an include name to its source, or null when the file is absent. */
export type WgslIncludeResolver = (name: string) => Promise<string | null>;

/** Cheap check so callers can skip the whole machinery for the 1417 files that have none. */
export function hasWgslInclude(source: string): boolean {
  return /^[ \t]*#include[ \t]/m.test(stripComments(source));
}

/**
 * Blanks out comment bodies while preserving every newline, so line numbers and
 * offsets stay aligned with the original. Used only to decide whether a
 * directive is live; the text that gets expanded is always the original.
 */
function stripComments(source: string): string {
  let out = '';
  let i = 0;
  while (i < source.length) {
    const two = source.slice(i, i + 2);
    if (two === '//') {
      while (i < source.length && source[i] !== '\n') {
        out += ' ';
        i += 1;
      }
    } else if (two === '/*') {
      // WGSL block comments nest.
      let depth = 0;
      while (i < source.length) {
        const pair = source.slice(i, i + 2);
        if (pair === '/*') {
          depth += 1;
          out += '  ';
          i += 2;
        } else if (pair === '*/') {
          depth -= 1;
          out += '  ';
          i += 2;
          if (depth === 0) break;
        } else {
          out += source[i] === '\n' ? '\n' : ' ';
          i += 1;
        }
      }
    } else {
      out += source[i];
      i += 1;
    }
  }
  return out;
}

function validateName(name: string, parent: string): void {
  if (name.includes('/') || name.includes('\\')) {
    throw new WgslIncludeError(
      `${parent}: #include "${name}" — public/shaders is flat, a path separator is not allowed`,
    );
  }
  if (!name.endsWith('.wgsl')) {
    throw new WgslIncludeError(`${parent}: #include "${name}" — must end in .wgsl`);
  }
  if (!name.startsWith(WGSL_LIBRARY_PREFIX)) {
    throw new WgslIncludeError(
      `${parent}: #include "${name}" — only "${WGSL_LIBRARY_PREFIX}"-prefixed library files may be included, ` +
        `so a catalog shader cannot be inlined into another`,
    );
  }
}

/**
 * Expands every `#include` in `source`.
 *
 * Returns the source unchanged when it contains no directive, so the
 * overwhelming majority of the catalog costs one regex test.
 *
 * @param source  WGSL text.
 * @param resolve Loads an include by name; returns null when it does not exist.
 * @param entry   Name used for the entry file in error messages and markers.
 */
export async function expandWgslIncludes(
  source: string,
  resolve: WgslIncludeResolver,
  entry = '<entry>',
): Promise<string> {
  if (!hasWgslInclude(source)) return source;

  // Every file included anywhere in this expansion, in include order. WGSL has
  // no include guards, so a second include of the same file is an error.
  const seen = new Map<string, string>();

  const expand = async (text: string, file: string, stack: string[]): Promise<string> => {
    if (stack.length > WGSL_INCLUDE_MAX_DEPTH) {
      throw new WgslIncludeError(
        `#include nesting deeper than ${WGSL_INCLUDE_MAX_DEPTH}: ${[...stack, file].join(' -> ')}`,
      );
    }

    const lines = text.split('\n');
    // Comment-blanked twin: same line count, so index i means the same line in both.
    const blanked = stripComments(text).split('\n');
    const out: string[] = [];

    for (let i = 0; i < lines.length; i += 1) {
      const match = blanked[i]?.match(INCLUDE_RE);
      if (!match) {
        out.push(lines[i]);
        continue;
      }

      const name = match[1];
      const lineNo = i + 1;
      validateName(name, file);

      if (stack.includes(name) || name === file) {
        throw new WgslIncludeError(
          `#include cycle: ${[...stack, file, name].join(' -> ')}`,
        );
      }
      const previous = seen.get(name);
      if (previous !== undefined) {
        throw new WgslIncludeError(
          `${file}:${lineNo}: #include "${name}" was already included by ${previous} — ` +
            `WGSL has no include guards, so the second copy would redeclare every symbol`,
        );
      }
      seen.set(name, `${file}:${lineNo}`);

      const included = await resolve(name);
      if (included === null) {
        throw new WgslIncludeError(`${file}:${lineNo}: #include "${name}" — file not found under public/shaders`);
      }

      out.push(`// #include-expanded: ${name} (from ${file}:${lineNo})`);
      out.push(await expand(included, name, [...stack, file]));
    }

    return out.join('\n');
  };

  return expand(source, entry, []);
}
